File: src/template/TemplateManager.php
<?php

namespace pocketcloud\cloud\template;

use pocketcloud\cloud\cache\impl\MaintenanceListCache;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\event\impl\template\TemplateCreateEvent;
use pocketcloud\cloud\event\impl\template\TemplateEditEvent;
use pocketcloud\cloud\event\impl\template\TemplateRemoveEvent;
use pocketcloud\cloud\group\ServerGroupManager;
use pocketcloud\cloud\network\packet\impl\TemplateSyncPacket;
use pocketcloud\cloud\player\CloudPlayer;
use pocketcloud\cloud\player\CloudPlayerManager;
use pocketcloud\cloud\provider\CloudProvider;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\server\prepare\ServerPreparator;
use pocketcloud\cloud\util\FileUtils;
use pocketcloud\cloud\util\misc\Loadable;
use pocketcloud\cloud\util\misc\Tickable;
use pocketcloud\cloud\util\trait\SingletonTrait;

final class TemplateManager implements Loadable, Tickable {
    use SingletonTrait;

    /** @var array<Template> */
    private array $templates = [];

    public function __construct() {
        self::setInstance($this);
    }

    public function load(): void {
        foreach (TemplateType::getAll() as $type) FileUtils::createDir($type->getGlobalTemplatePath());

        CloudProvider::current()->getTemplates()
            ->then(function(array $templates): void {
                $this->templates = $templates;

                if (array_sum(array_map(fn(Template $template) => $template->getMinServerCount(), array_filter($this->templates, fn(Template $template) => $template->isAutoStart()))) >= 9 && count(ServerPreparator::getInstance()->getThreads()) == 0) {
                    CloudLogger::get()->warn("Your total active server count exceeds §b9§8, §rtherefore you should set §8'§bserverPrepareThreads§8' §rinside your §bconfig.json §rto at least §b1 §ror §b2 §rand restart the the §bcloud§r.");
                }

                ServerGroupManager::getInstance()->load();
            });
    }

    public function create(Template $template): void {
        $startTime = microtime(true);
        CloudProvider::current()->addTemplate($template);

        new TemplateCreateEvent($template)->call();

        CloudLogger::get()->debug("Creating directory: " . $template->getPath());
        if (!file_exists($template->getPath())) mkdir($template->getPath());
        $this->templates[$template->getName()] = $template;
        CloudLogger::get()->success("Successfully §acreated §rthe template §b" . $template->getName() . "§r. §8(§rTook §b" . number_format(microtime(true) - $startTime, 3) . "s§8)");
        TemplateSyncPacket::create($template, false)->broadcastPacket();
    }

    public function remove(Template $template): void {
        $startTime = microtime(true);
        CloudProvider::current()->removeTemplate($template);

        new TemplateRemoveEvent($template)->call();

        CloudServerManager::getInstance()->stop($template, true);

        if (file_exists($template->getPath())) FileUtils::removeDirectory($template->getPath());
        if (isset($this->templates[$template->getName()])) unset($this->templates[$template->getName()]);
        CloudLogger::get()->success("Successfully §cremoved §rthe template §b" . $template->getName() . "§r. §8(§rTook §b" . number_format(microtime(true) - $startTime, 3) . "s§8)");
        TemplateSyncPacket::create($template, true)->broadcastPacket();
    }

    public function edit(Template $template, ?bool $lobby, ?bool $maintenance, ?bool $static, ?bool $alwaysCopyToStaticServers, ?int $maxPlayerCount, ?int $minServerCount, ?int $maxServerCount, ?float $startNewPercentage, ?bool $autoStart): void {
        $startTime = microtime(true);
        $template->setLobby($lobby === null ? $template->isLobby() : $lobby);
        $template->setMaintenance($maintenance === null ? $template->isMaintenance() : $maintenance);
        $template->setStatic($static === null ? $template->isStatic() : $static);
        $template->setAlwaysCopyToStaticServers($alwaysCopyToStaticServers === null ? $template->isAlwaysCopyToStaticServers() : $alwaysCopyToStaticServers);
        $template->setMaxPlayerCount($maxPlayerCount === null ? $template->getMaxPlayerCount() : $maxPlayerCount);
        $template->setMinServerCount($minServerCount === null ? $template->getMinServerCount() : $minServerCount);
        $template->setMaxServerCount($maxServerCount === null ? $template->getMaxServerCount() : $maxServerCount);
        $template->setStartNewPercentage($startNewPercentage === null ? $template->getStartNewPercentage() : $startNewPercentage);
        $template->setAutoStart($autoStart === null ? $template->isAutoStart() : $autoStart);

        new TemplateEditEvent($template, $lobby, $maintenance, $static, $maxPlayerCount, $minServerCount, $maxServerCount, $startNewPercentage, $autoStart)->call();

        CloudProvider::current()->editTemplate($template, $template->write());

        CloudLogger::get()->success("Successfully §eedited §rthe template §b" . $template->getName() . "§r. §8(§rTook §b" . number_format(microtime(true) - $startTime, 3) . "s§8)");
        TemplateSyncPacket::create($template, false)->broadcastPacket();

        if ($template->isMaintenance()) {
            foreach (array_filter(CloudPlayerManager::getInstance()->getAll($template), fn (CloudPlayer $player): bool => !MaintenanceListCache::is($player->getName())) as $player) {
                $player->kick("MAINTENANCE");
            }
        }
    }

    public function check(string $name): bool {
        return isset($this->templates[$name]);
    }

    public function tick(int $currentTick): void {
        if (!ServerGroupManager::getInstance()->isLoaded()) return;
        foreach (TemplateManager::getInstance()->getAll() as $template) {
            if ($template->isAutoStart()) {
                if (($running = count(CloudServerManager::getInstance()->getAll($template))) < $template->getMaxServerCount()) {
                    if ((microtime(true) - CloudServerManager::getInstance()->getLastServerStopTime()) >= 0.5)
                        CloudServerManager::getInstance()->start($template, ($template->getMinServerCount() - $running));
                }
            }

            if (($latest = CloudServerManager::getInstance()->getLatest($template)) !== null) {
                $players = $latest->getPlayerCount();
                $requiredPercentage = $template->getStartNewPercentage();
                if ($requiredPercentage <= 0) continue;
                $percentage = (100 * $players) / $latest->getServerData()->getMaxPlayers();
                if ($percentage >= $requiredPercentage && CloudServerManager::getInstance()->checkCapacity($template)) {
                    CloudServerManager::getInstance()->start($template);
                }
            }
        }
    }

    public function get(string $name): ?Template {
        return $this->templates[$name] ?? null;
    }

    public function getAll(TemplateType ...$type): array {
        if (count($type) > 0) return array_filter($this->templates, fn(Template $template) => array_any($type, fn(TemplateType $type) => $template->getTemplateType()->getName() == $type->getName()));
        return $this->templates;
    }
}

File: src/main/java/de/pocketcloud/cloud/template/TemplateManager.java
package de.pocketcloud.cloud.template;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.event.impl.template.TemplateCreateEvent;
import de.pocketcloud.cloud.event.impl.template.TemplateEditEvent;
import de.pocketcloud.cloud.event.impl.template.TemplateRemoveEvent;
import de.pocketcloud.cloud.load.Loadable;
import de.pocketcloud.cloud.provider.CloudProvider;
import de.pocketcloud.cloud.template.group.ServerGroupManager;
import de.pocketcloud.cloud.template.util.TemplateEditData;
import de.pocketcloud.cloud.tick.Tickable;
import de.pocketcloud.cloud.util.FileUtils;
import de.pocketcloud.cloud.util.Utils;
import de.pocketcloud.cloud.util.benchmark.Benchmark;
import de.pocketcloud.cloud.util.benchmark.BenchmarkTiming;
import lombok.Getter;

import java.io.IOException;
import java.nio.file.Files;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class TemplateManager implements Tickable, Loadable {

    @Getter
    private static TemplateManager instance = null;

    private final Map<String, Template> templates = new HashMap<>();

    public TemplateManager() {
        instance = this;
    }

    //TODO broadcast sync packets

    @Override
    public void load() {
        CloudLogger.get().info("Loading templates...");
        for (TemplateType type : TemplateType.values()) FileUtils.createDir(type.globalTemplatePath());
        CloudProvider.current().getTemplates()
                .thenAccept(this.templates::putAll)
                .thenAccept(_ -> ServerGroupManager.getInstance().load());
    }

    @Override
    public void unload() {
        CloudLogger.get().info("Unloading templates...");
        this.templates.clear();
        ServerGroupManager.getInstance().unload();
    }

    public void create(Template template) {
        Benchmark.startTiming("template_creation");
        try {
            TemplateCreateEvent ev = new TemplateCreateEvent(template);
            ev.call();
            if (ev.isCancelled()) {
                Benchmark.stopTiming("template_creation");
                return;
            }

            CloudProvider.current().addTemplate(template);
            if (!Files.isDirectory(template.path())) Files.createDirectories(template.path());
            templates.put(template.name(), template);
            BenchmarkTiming res = Benchmark.stopTiming("template_creation");
            CloudLogger.get().success("Successfully §acreated §rthe template §b{}§r. §8(§rTook §b{}ms§8)", template.name(), Utils.formatNumber(res.duration(), 2));
        } catch (IOException e) {
            CloudLogger.get().exception("Failed to create template {}", e, template.name());
            Benchmark.stopTiming("template_creation");
            templates.remove(template.name());
        }
    }

    public void edit(Template template, TemplateEditData editData) {
        Benchmark.startTiming("template_editing");
        TemplateEditEvent ev = new TemplateEditEvent(template, editData);
        ev.call();
        if (ev.isCancelled()) {
            Benchmark.stopTiming("template_editing");
            return;
        }

        editData.applyTo(template);
        CloudProvider.current().editTemplate(template, template.write());
        BenchmarkTiming res = Benchmark.stopTiming("template_editing");
        CloudLogger.get().success("Successfully §eedited §rthe template §b{}§r. §8(§rTook §b{}ms§8)", template.name(), Utils.formatNumber(res.duration(), 2));
        //TODO kick playxers from template if new edit data has maintenance=true
    }

    public void remove(Template template) {
        Benchmark.startTiming("template_removal");
        TemplateRemoveEvent ev = new TemplateRemoveEvent(template);
        ev.call();
        if (ev.isCancelled()) {
            Benchmark.stopTiming("template_removal");
            return;
        }

        CloudProvider.current().removeTemplate(template);
        if (Files.isDirectory(template.path())) {
            try {
                FileUtils.removeDirectory(template.path());
            } catch (Exception e) {
                CloudLogger.get().exception("Failed to remove template directory of {}", e, template.name());
            }
        }

        templates.remove(template.name());
        BenchmarkTiming res = Benchmark.stopTiming("template_removal");
        CloudLogger.get().success("Successfully §cremoved §rthe template §b{}§r. §8(§rTook §b{}ms§8)", template.name(), Utils.formatNumber(res.duration(), 2));
    }

    public boolean check(String name) {
        return templates.containsKey(name);
    }

    @Override
    public void tick(long currentTick) {
        //start servers
    }

    public Optional<Template> get(String name) {
        return Optional.ofNullable(templates.getOrDefault(name, null));
    }

    public List<Template> getAll(TemplateType... types) {
        if (types.length > 0) return templates.values().stream()
                .filter(template -> {
                    for (TemplateType type : types) {
                        if (template.isTypeOf(type)) return true;
                    }
                    return false;
                }).toList();
        return templates.values().stream().toList();
    }
}